src/java.base/share/classes/jdk/internal/foreign/LayoutRecordMapper.java

```diff
@@ -42,301 +42,289 @@
 import java.lang.reflect.RecordComponent;
 import java.util.ArrayList;
 import java.util.Arrays;
-import java.util.Collections;
-import java.util.LinkedHashSet;
 import java.util.List;
 import java.util.Objects;
-import java.util.Set;
-import java.util.function.BinaryOperator;
+import java.util.Optional;
 import java.util.function.Function;
 import java.util.stream.Collectors;
-import java.util.stream.IntStream;
 import java.util.stream.Stream;
 
-import static java.util.stream.Collectors.toMap;
-
 /**
  * A record mapper that is matching a GroupLayout to match the components of a record.
  *
  * @param <T> the Record type
  */
-public final class LayoutRecordMapper<T extends Record>
+public final class LayoutRecordMapper<T>
         implements Function<MemorySegment, T> {
 
-    enum Allow {EXACT, BOXING, BOXING_NARROWING_AND_WIDENING}
-
-    public static final Allow ALLOW = Allow.EXACT;
-
     private static final MethodHandles.Lookup LOOKUP = MethodHandles.lookup();
     private static final MethodHandles.Lookup PUBLIC_LOOKUP = MethodHandles.publicLookup();
 
     private final Class<T> type;
     private final GroupLayout layout;
+    private final long offset;
+    private final int depth;
+    private final MethodHandles.Lookup lookup;
     private final MethodHandle ctor;
 
     public LayoutRecordMapper(Class<T> type,
                               GroupLayout layout) {
-        this(type, layout, 0);
+        this(type, layout, 0L, 0, PUBLIC_LOOKUP);
```

Comment:
I tend to agree with Remi that passing an explicit lookup to the mapper factory is unavoidable here.